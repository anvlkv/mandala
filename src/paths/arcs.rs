use crate::{Angle, GlVec, Point, Vector, VectorValuedFn};

/// sweeps an arc of radius with center, start and sweep angles
#[derive(Debug, Clone, Copy)]
pub struct SweepArc {
    pub radius: Vector,
    pub center: Point,
    pub start_angle: Angle,
    pub sweep_angle: Angle,
}

impl Default for SweepArc {
    fn default() -> Self {
        Self {
            radius: GlVec::default().into(),
            center: GlVec::default().into(),
            start_angle: Angle::default(),
            sweep_angle: Angle::default(),
        }
    }
}

impl VectorValuedFn for SweepArc {
    fn eval(&self, t: crate::Float) -> Vector {
        let angle = self.start_angle + self.sweep_angle * t;

        crate::Vector {
            x: self.center.x + self.radius.x * angle.cos(),
            y: self.center.y + self.radius.y * angle.sin(),
            #[cfg(feature = "3d")]
            z: self.center.z + self.radius.z * angle.sin(),
        }
    }

    fn length(&self) -> crate::Float {
        self.radius.x.hypot(self.radius.y) * self.sweep_angle.to_radians()
    }
}

/// draws an arc between two points
#[derive(Debug, Clone, Copy)]
pub struct ArcSegment {
    pub start: Point,
    pub end: Point,
    pub radius: Vector,
    /// draws largest of two arcs
    pub large_arc: bool,
    /// draws arc in the direction of increasing angle
    pub poz_angle: bool,
}

impl Default for ArcSegment {
    fn default() -> Self {
        Self {
            start: GlVec::default().into(),
            end: GlVec::default().into(),
            radius: GlVec::default().into(),
            large_arc: false,
            poz_angle: false,
        }
    }
}

impl ArcSegment {
    /// finds center of the arc based on `large_arc`
    /// and `poz_angle` flags
    pub fn arc_center(&self) -> Point {
        let mid_point = crate::Vector {
            x: (self.start.x + self.end.x) / 2.0,
            y: (self.start.y + self.end.y) / 2.0,
            #[cfg(feature = "3d")]
            z: (self.start.z + self.end.z) / 2.0,
        };

        let start_to_end: GlVec = crate::Vector {
            x: self.end.x - self.start.x,
            y: self.end.y - self.start.y,
            #[cfg(feature = "3d")]
            z: self.end.z - self.start.z,
        }
        .into();

        let start_to_mid: GlVec = crate::Vector {
            x: mid_point.x - self.start.x,
            y: mid_point.y - self.start.y,
            #[cfg(feature = "3d")]
            z: mid_point.z - self.start.z,
        }
        .into();

        let mut angle = start_to_end.angle_between(start_to_mid);

        if self.large_arc {
            angle = if self.poz_angle { angle } else { -angle };
        } else {
            angle = if self.poz_angle { -angle } else { angle };
        }

        let center_x = self.start.x + self.radius.x * angle.cos();
        let center_y = self.start.y + self.radius.y * angle.sin();
        #[cfg(feature = "3d")]
        let center_z = self.start.z + self.radius.z * angle.sin();

        crate::Vector {
            x: center_x,
            y: center_y,
            #[cfg(feature = "3d")]
            z: center_z,
        }
        .into()
    }
}

impl VectorValuedFn for ArcSegment {
    fn eval(&self, t: crate::Float) -> Vector {
        let center = self.arc_center();
        let start_angle = Angle::from_radians(
            (GlVec::from(self.end) - GlVec::from(self.start))
                .angle_between(GlVec::from(self.radius)),
        );

        let sweep_angle = if self.large_arc {
            Angle::PI
        } else {
            Angle::FRAC_PI_2
        };

        let angle = start_angle + sweep_angle * t;

        crate::Vector {
            x: center.x + self.radius.x * angle.cos(),
            y: center.y + self.radius.y * angle.sin(),
            #[cfg(feature = "3d")]
            z: center.z + self.radius.z * angle.sin(),
        }
    }

    fn length(&self) -> crate::Float {
        let start_angle = Angle::from_radians(
            (GlVec::from(self.end) - GlVec::from(self.start))
                .angle_between(GlVec::from(self.radius)),
        );

        let sweep_angle = if self.large_arc {
            Angle::PI
        } else {
            Angle::FRAC_PI_2
        };

        (start_angle + sweep_angle).to_radians() * self.radius.x.hypot(self.radius.y)
    }
}

#[cfg(test)]
mod arc_tests {
    use super::*;
    use crate::test_util::test_name;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_sweep_arc() {
        let arc = SweepArc {
            radius: Vector {
                x: 10.0,
                y: 10.0,
                #[cfg(feature = "3d")]
                z: 0.0,
            },
            center: Point {
                x: 0.0,
                y: 0.0,
                #[cfg(feature = "3d")]
                z: 0.0,
            },
            start_angle: Angle::from_degrees(0.0),
            sweep_angle: Angle::from_degrees(90.0),
        };
        let points: Vec<_> = arc.sample_evenly(10);
        assert_debug_snapshot!(test_name("sweep-arc"), points);
    }

    #[test]
    fn test_arc_segment() {
        let arc = ArcSegment {
            start: Point {
                x: 0.0,
                y: 10.0,
                #[cfg(feature = "3d")]
                z: 0.0,
            },
            end: Point {
                x: 10.0,
                y: 0.0,
                #[cfg(feature = "3d")]
                z: 0.0,
            },
            radius: Vector {
                x: 10.0,
                y: 10.0,
                #[cfg(feature = "3d")]
                z: 0.0,
            },
            large_arc: true,
            poz_angle: true,
        };
        let points: Vec<_> = arc.sample_evenly(10);
        assert_debug_snapshot!(test_name("segment-arc"), points);
    }
}
