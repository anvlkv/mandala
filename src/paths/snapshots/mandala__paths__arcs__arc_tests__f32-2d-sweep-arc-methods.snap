---
source: src/paths/arcs.rs
expression: "(eval_points, sample_points, derivative_points, normal_points)"
---
(
    [
        Vector2 {
            x: 10.0,
            y: 0.0,
        },
        Vector2 {
            x: 9.8768835,
            y: 1.5643448,
        },
        Vector2 {
            x: 9.510566,
            y: 3.09017,
        },
        Vector2 {
            x: 8.910066,
            y: 4.5399055,
        },
        Vector2 {
            x: 8.09017,
            y: 5.8778524,
        },
        Vector2 {
            x: 7.071068,
            y: 7.071068,
        },
        Vector2 {
            x: 5.877852,
            y: 8.090171,
        },
        Vector2 {
            x: 4.539905,
            y: 8.910066,
        },
        Vector2 {
            x: 3.0901697,
            y: 9.510566,
        },
        Vector2 {
            x: 1.5643449,
            y: 9.8768835,
        },
        Vector2 {
            x: -4.371139e-7,
            y: 10.0,
        },
    ],
    [
        Vector2 {
            x: 10.0,
            y: 0.0,
        },
        Vector2 {
            x: 9.848078,
            y: 1.7364819,
        },
        Vector2 {
            x: 9.396926,
            y: 3.4202015,
        },
        Vector2 {
            x: 8.660254,
            y: 5.0,
        },
        Vector2 {
            x: 7.6604443,
            y: 6.4278765,
        },
        Vector2 {
            x: 6.4278755,
            y: 7.660445,
        },
        Vector2 {
            x: 4.9999995,
            y: 8.6602545,
        },
        Vector2 {
            x: 3.4202003,
            y: 9.396927,
        },
        Vector2 {
            x: 1.7364811,
            y: 9.848078,
        },
        Vector2 {
            x: -4.371139e-7,
            y: 10.0,
        },
    ],
    [
        Vector2 {
            x: 0.0,
            y: 15.702821,
        },
        Vector2 {
            x: -2.45692,
            y: 15.514572,
        },
        Vector2 {
            x: -4.854454,
            y: 14.939358,
        },
        Vector2 {
            x: -7.130454,
            y: 13.995743,
        },
        Vector2 {
            x: -9.232439,
            y: 12.707207,
        },
        Vector2 {
            x: -11.107239,
            y: 11.107239,
        },
        Vector2 {
            x: -12.709278,
            y: 9.23382,
        },
        Vector2 {
            x: -13.995743,
            y: 7.130454,
        },
        Vector2 {
            x: -14.940739,
            y: 4.854454,
        },
        Vector2 {
            x: -15.516298,
            y: 2.45692,
        },
        Vector2 {
            x: -15.707922,
            y: 0.0,
        },
    ],
    [
        Vector2 {
            x: -1.0,
            y: 0.0,
        },
        Vector2 {
            x: -0.9876918,
            y: -0.15641293,
        },
        Vector2 {
            x: -0.9510497,
            y: -0.30903786,
        },
        Vector2 {
            x: -0.8910254,
            y: -0.45395347,
        },
        Vector2 {
            x: -0.80901355,
            y: -0.58778995,
        },
        Vector2 {
            x: -0.7071068,
            y: -0.7071068,
        },
        Vector2 {
            x: -0.58778477,
            y: -0.8090173,
        },
        Vector2 {
            x: -0.45395347,
            y: -0.8910254,
        },
        Vector2 {
            x: -0.30901203,
            y: -0.95105815,
        },
        Vector2 {
            x: -0.15639594,
            y: -0.98769444,
        },
        Vector2 {
            x: -0.0,
            y: -0.99999994,
        },
    ],
)
